doc/dev/osd_internals/multi_object_write_transaction.rst

```diff
@@ -0,0 +1,117 @@
+     itself has released transaction context
+
+  #. *MASTER* collects response from *SLAVE*, if all *EINPROGRESS*, then sends
+     *SLAVE* *LOG* message, the transaction enters the *LOG* state
+     
+  #. *SLAVE* receives *LOG*, constructs a write op which persists the 
+     multi-object operation information in the object attribute named 
+     *MOWO_ATTR*, and also synchronizes the op within its replica osds, 
+     utilizing the present single object write transaction mechanism, then 
+     returns *MASTER* zero
+     
+  #. *MASTER* collects response from *SLAVE*, if all zero, persists the 
+     multi-object operation information itself, then performs the write op on 
+     the master object, and synchronizes the op within its replica osds, return 
+     client *ACK*
+     
+  #. *MASTER* sends *SLAVE* *COMMIT & UNLOCK* message, the transaction enters 
+     the *COMMIT* state
+     
+  #. *SLAVE* receives *COMMIT & UNLOCK* message, first processes *COMMIT*, 
+     performs the write op on slave object, and synchronizes the op within its
+     replica osds, returns *MASTER* zero, *SLAVE* processes *UNLOCK*, cleans up 
+     the transaction and deletes the persisted object attribute
+     
+  #. *MASTER* collects response from *SLAVE*, if all zero, clean up the 
+     transaction and deletes the persisted object attribute, acks *SLAVE* 
+     *QUERY* message
+     
+     
+Fault Tolerance
+---------------
+
+- If *SLAVE* down at any time before persisting transaction to object 
+  attribute, when *MASTER* resends *PREPARE* or (re)sends *LOG*, since *SLAVE* 
+  has lost the in-memory transaction information, it will return *EPERM*. If 
+  *MASTER* receives *EPERM*, it recognizes the situation and redo *PREPARE* for 
+  this *SLAVE* to resend the operation to it to proceed
+
+- If *MASTER* down at any time before persisting transaction to object 
+  attribute, when client resends the request, it will reconstruct the 
+  transaction, and resends *PREPARE* and *LOG* to *SLAVE*, if *SLAVE* is alive, 
+  it will recognize these resent messages, does nothing and replies with 
+  correct response to proceed; If *SLAVE* also restarts (for example, share the 
+  same osd with *MASTER*), it will return *EPERM* and *MASTER* will resends 
+  *PREPARE* which falls back to the above situation
+
+- If *SLAVE* down at any time after persisting transaction to object attribute, 
+  when *MASTER* resends *LOG*/sends *COMMIT*, *SLAVE* will reconstruct the 
+  transaction from the object attribute and proceed. If *SLAVE* down during 
+  *UNLOCK*, when *MASTER* finishes transaction, it will ack *QUERY* to the 
+  *SLAVE* to release the transaction to proceed
+
+- If *MASTER* down at any time after persisting transaction to object 
+  attribute, when client resends request/*SLAVE* resends *QUERY*, *MASTER* will 
+  reconstruct the transaction from the object attribute and send *COMMIT* to 
+  *SLAVE* to proceed
+
```

Comment:
Clients don't undergo interval changes, they just die.  The slave can't be continuously sending queries to the master, something would have to trigger it.  It also can't merely do it when it gets an IO on the object -- it would either be redundant or inadequate (what if the master dies after the blocked IO arrives but before it completes the transaction?)  Does the slave monitor osdmap changes to detect an interval change on the primary PG for all operations for which it is the slave?  If so, that seems like it duplicates a great deal of the Objecter.  Does it poll?  If so, at what interval?